Fix Faker being reset for every registered call and after create

circulon opened the pull request:
This Fixes the faker instance producing duplicate data when using .unique
This was due to a new Faker instance being created every time a factory callback is used

Example being if a factory is used to create multiple  users that each required a unique email address, the same email address would be generated every time even if faker.unique.ascii_email is used

circulon committed:
Fix Faker being reset for every registered call and after create

This Fixes the faker instance producing duplicate data when using .unique
This was due to a new Faker instance being created every time a factory callback is used

josephmancuso commented:
why don't you want a new instance of Faker? Whats the benefit of having a singleton style faker instance?

circulon commented:
> why don't you want a new instance of Faker? Whats the benefit of having a singleton style faker instance?

Ok lets take this scenario:
a company has a unique email address (email has a UNIQUE Index on the column)
Each user in that company also has a unique company email address (email has a UNIQUE Index on the column).
This is a fairly standard data scenario.

For a Faker instance to generate unique data it MUST be seeded, which it is not by default, as such it will generate the same data for .unique calls. 

In its current state the following factories will fail with a duplicate email error on the second user of the first company as the same name is generated for the new instance of faker applied to each user model in after_create.

```python
def company_factory(faker):
    return {
        "name": faker.unique.company(),
        "email": faker.unique.ascii_company_email(),
        "phone": faker.unique.phone_number(),
    }

def user_factory(faker):
    return {}

def after_company(model, faker):
    company_email = model.email.split(sep="@")

    user_count = faker.random_int(min=2, max=25)
    for i in range(user_count):
        user_name = faker.unique.name().split()
        user_email = "{}.{}@{}".format(user_name[0], user_name[1], company_email[1]).lower().replace("..", ".")
        result = User.create({
            "email": user_email,
            'company_id': model.id,
            "first_name": user_name[0],
            "last_name": user_name[1],
            "phone": faker.unique.phone_number(),
           })

Factory.register(Company, company_factory)
Factory.register(User, user_factory)
Factory.after_creating(Company, after_company)

Factory(Company, 10).create()

```

By using the Faker instance in a singletome type manner the .unique(ness) of all values is ensured across the entire seed:run

@josephmancuso 
I hope this clarifies things

josephmancuso commented:
@circulon got it. That makes a lot of sense actually

josephmancuso committed:
Merge branch '1.0' into patch/fix_seed_faker_not_unique

circulon commented:
Hi @josephmancuso 

Would it be possible to get these changes merged and a new release tagged please?

josephmancuso commented:
@circulon sure I'll make sure I bump priority on this

josephmancuso commented:
@circulon this is in 1.0.57
